R/Module-StrategusModule.R

```diff
@@ -0,0 +1,309 @@
+    #' Each module will have its own implementation if it needs to create
+    #' a shared resource.
+    #' @param className The class name of the shared resources specifications
+    #' @param sharedResourcesSpecifications The shared resources specifications
+    createSharedResourcesSpecifications = function(className, sharedResourcesSpecifications) {
+      class(sharedResourcesSpecifications) <- c(className, self$internalSharedResourcesClassName)
+      return(sharedResourcesSpecifications)
+    },
+    #' @description Base function for validating the module settings object.
+    #' Each module will have its own implementation and this base class method
+    #' will be used to ensure the module specifications are valid ahead of
+    #' execution
+    #' @template moduleSpecifications
+    validateModuleSpecifications = function(moduleSpecifications) {
+      errorMessages <- checkmate::makeAssertCollection()
+      checkmate::assertClass(moduleSpecifications, self$internalModuleSpecificationClassName)
+      checkmate::assertClass(moduleSpecifications, self$moduleClassName)
+      checkmate::reportAssertions(collection = errorMessages)
+    },
+    #' @description Base function for validating the shared resources
+    #' specification settings object. Each module will have its own
+    #' implementation and this base class method will be used to ensure
+    #' the module specifications are valid ahead of execution
+    #' @param className The class name of the shared resources specifications
+    #' @param sharedResourcesSpecifications The shared resources specifications
+    validateSharedResourcesSpecifications = function(className, sharedResourcesSpecifications) {
+      errorMessages <- checkmate::makeAssertCollection()
+      checkmate::assertClass(sharedResourcesSpecifications, self$internalSharedResourcesClassName)
+      checkmate::assertClass(sharedResourcesSpecifications, className)
+      checkmate::reportAssertions(collection = errorMessages)
+    }
+  ),
+  private = list(
+    jobContext = JobContext$new(),
+    .message = function(...) {
+      rlang::inform(paste0(...))
+    },
+    .createLoggers = function(moduleExecutionSettings) {
+      # Establish loggers for the module execution
+      if (!dir.exists(moduleExecutionSettings$resultsSubFolder)) {
+        dir.create(moduleExecutionSettings$resultsSubFolder, recursive = T)
+      }
+      ParallelLogger::addDefaultFileLogger(
+        name = "MODULE_LOGGER",
+        fileName = file.path(moduleExecutionSettings$resultsSubFolder, "log.txt")
+      )
+      ParallelLogger::addDefaultErrorReportLogger(
+        name = "MODULE_ERROR_LOGGER",
+        file.path(moduleExecutionSettings$resultsSubFolder, "errorReportR.txt")
+      )
+    },
+    .clearLoggers = function() {
+      ParallelLogger::unregisterLogger("MODULE_LOGGER")
+      ParallelLogger::unregisterLogger("MODULE_ERROR_LOGGER")
+    },
```

Comment:
Making a note based on discussions with @chrisknoll that we may want to remove this code for logging from inside of the module to the calling code. Right now, the `Strategus::execute` function allows for the specification of a log file for the entire execution of an analysis specification which feels sufficient.

Reply:
I think the core of the problem is that if you call execute and validate (or execute and upload) together, it seems to try to call `.createLoggers` twice, but yes, I'm hoping we can revisit how logging should work in the module context.

Reply:
I'm removing logging from within the modules and agree that it is cleaner for the calling process to attach the logging where desired.